File: Python/chess.py
################################################################################
# N queens problem (Python version)
# Class Chess
#
# Copyleft 2013 Vikman - All rights revoked.
# vikman90.blogspot.com - vmfdez90@gmail.com
# February 8, 2013
#
################################################################################

from copy import copy
from time import clock
from sys import argv
from random import seed, randint

seed()

class Chess:
    def __init__(self, size):
        '''Constructor'''

        self._size = size               # Number of queens
        # Queens' positions (candidates)
        self._queens = [[1 for j in range(size)] for i in range(size)]
        # Number of available values in each row
        self._queensCount = [size for i in range(size)]
        self._nsteps = 0                # Calls to assign()
        self._ndiscards = 0             # Calls to discard()
        self._stackDiscarded = list()   # Discarded candidates (index-value)
        self._stackCount = list()       # Discards in the last assignation

    def solve(self):
        '''Solve queens problem'''

        index = self.selectIndex()

        if index == -1:
            return True

        row = copy(self._queens[index])
        values = self.selectValues(row)

        for value in values:
            if not self.assign(index, value):
                self._queens[index] = copy(row)
                self._queensCount[index] = len(values)
                continue

            if self.solve():
                return True

            self.restoreLast()
            self._queens[index] = copy(row)
            self._queensCount[index] = len(values)

        return False

    def selectIndex(self):
        '''Get index of a unsolved row (minimum remaining values heuristics)'''
        minSize = self._size + 1
        index = -1

        for i in range(self._size):
            curSize = self._queensCount[i]

            if curSize > 1 and curSize < minSize:
                index = i
                minSize = curSize

        return index

    def assign(self, index, value):
        '''Assign a value to a row and propagate constraints'''

        self._queens[index] = [0 for i in range(self._size)]
        self._queensCount[index] = 0
        self._stackCount.append(0)
        self._nsteps += 1

        for i in range(self._size):
            if i == index:
                continue

            diag1 = value + (index - i)
            diag2 = value - (index - i)

            if not (self.discard(i, value) and self.discard(i, diag1)
                    and self.discard(i, diag2)):
                self.restoreLast()
                return False

        self._queens[index][value] = 1
        self._queensCount[index] = 1
        return True

    def discard(self, index, value):
        '''Discard candidate values (constraints propagation)'''

        if value < 0 or value >= self._size or not self._queens[index][value]:
            return True

        self._ndiscards += 1
        self._queens[index][value] = 0
        self._queensCount[index] -= 1
        self._stackDiscarded.append([index, value])
        self._stackCount.append(self._stackCount.pop() + 1)

        if self._queensCount[index] == 0:
            return False

        if self._queensCount[index] == 1:
            value = self.getValue(index)

            for i in range(self._size):
                if i == index:
                    continue

                diag1 = value + (index - i)
                diag2 = value - (index - i)

                if not (self.discard(i, value) and self.discard(i, diag1)
                        and self.discard(i, diag2)):
                    return False

        return True

    def restoreLast(self):
        '''Undo last assignation (restore constraints)'''

        count = self._stackCount.pop()

        for i in range(count):
            pair = self._stackDiscarded.pop()

            if not self._queens[pair[0]][pair[1]]:
                self._queens[pair[0]][pair[1]] = 1
                self._queensCount[pair[0]] += 1

    def steps(self):
        '''Get number of total tries of assignation'''

        return self._nsteps

    def discards(self):
        '''Get number of total discards'''

        return self._ndiscards

    def selectValues(self, row):
        '''Select all available indices from a row'''

        values = []
        offset = randint(0, self._size - 1)

        for i in range(len(row)):
            index = (offset + i) % self._size

            if row[index]:
                values.append(index)

        return values

    def getValue(self, index):
        '''Get the only value that is set in the indexed row'''

        for i in range(self._size):
            if self._queens[index][i]:
                return i

        return -1

    def __str__(self):
        '''Representation of the chessboard'''

        string = str()

        for i in range(self._size):
            if self._queensCount[i] == 0:
                string += "Queen " + str(i + 1) + " not solved.\n"
            else:
                value = self._queens[i][0] + 1
                string += "Queen " + str(i + 1) + ": square " + str(value) + "\n"

        return string[:-1]

File: JS/chess.js
/*******************************************************************************
 * N queens problem (JavaScript version)
 * Class Chess
 *
 * Copyleft 2016 Vikman - All rights revoked.
 * vikman90.blogspot.com - vmfdez90@gmail.com
 * March 26, 2016
 *
 ******************************************************************************/

function fill(length, value) {
    var array = new Array(value)

    for (var i = 0; i < length; i++)
        array[i] = value

    return array
}

function Chess(size) {

    //--------------------------------------------------------------------------
    // Constructor

    var complete = fill(size, 1)

    this.size = size
    this.queens = new Array(size)
    this.queensCount = fill(size, size)
    this.nsteps = 0
    this.ndiscards = 0
    this.stackDiscarded = []
    this.stackCount = []

    for (var i = 0; i < size; i++)
        this.queens[i] = complete.slice()

    //--------------------------------------------------------------------------
    // Solve queens problem

    this.solve = function() {
        var index = this.selectIndex()

        if (index == -1)
            return true

        var row = this.queens[index].slice()
        var values = this.selectValues(row)

        for (var i in values) {
            var value = values[i]

            if (!this.assign(index, value)) {
                this.queens[index] = row.slice()
                this.queensCount[index] = values.length
                continue
            }

            if (this.solve())
                return true

            this.restoreLast()
            this.queens[index] = row.slice()
            this.queensCount[index] = values.length
        }

        return false
    }

    //--------------------------------------------------------------------------
    // Get index of a unsolved row (minimum remaining values heuristics)

    this.selectIndex = function() {
        var minsize = this.size + 1
        var index = -1

        for (var i = 0; i < this.size; i++) {
            var cursize = this.queensCount[i]

            if (cursize > 1 && cursize < minsize) {
                index = i
                minsize = cursize
            }
        }

        return index
    }

    //--------------------------------------------------------------------------
    // Assign a value to a row and propagate constraints

    this.assign = function(index, value) {
        this.queens[index] = fill(this.size, 0)
        this.queensCount[index] = 0
        this.stackCount.push(0)
        this.nsteps++

        for (var i = 0; i < this.size; i++) {
            if (i == index)
                continue

            var diag1 = value + (index - i)
            var diag2 = value - (index - i)

            if (!(this.discard(i, value) && this.discard(i, diag1) && this.discard(i, diag2))) {
                this.restoreLast()
                return false
            }
        }

        this.queens[index][value] = 1
        this.queensCount[index] = 1
        return true
    }

    //--------------------------------------------------------------------------
    // Discard candidate values (constraints propagation)

    this.discard = function(index, value) {
        if (value < 0 || value >= this.size || !this.queens[index][value])
            return true

        this.ndiscards++
        this.queens[index][value] = 0
        this.queensCount[index]--
        this.stackDiscarded.push([index, value])
        this.stackCount[this.stackCount.length - 1]++

        if (this.queensCount[index] == 0)
            return false

        if (this.queensCount[index] == 1) {
            value = this.queens[index].indexOf(1)

            for (var i = 0; i < this.size; i++) {
                if (i == index)
                    continue

                var diag1 = value + (index - i)
                var diag2 = value - (index - i)

                if (!(this.discard(i, value) && this.discard(i, diag1) && this.discard(i, diag2)))
                    return false
            }
        }

        return true
    }

    //--------------------------------------------------------------------------
    // Undo last assignation (restore constraints)

    this.restoreLast = function() {
        var n = this.stackCount.pop()

        for (var i = 0; i < n; i++) {
            var pair = this.stackDiscarded.pop()

            if (!this.queens[pair[0]][pair[1]]) {
                this.queens[pair[0]][pair[1]] = 1
                this.queensCount[pair[0]]++
            }
        }
    }

    //--------------------------------------------------------------------------
    // Select all available indices from a row

    this.selectValues = function(row) {
        var values = []
        var offset = parseInt(Math.random() * this.size)

        for (var i = 0; i < row.length; i++) {
            var index = (offset + i) % this.size

            if (row[index])
                values.push(index)
        }

        return values
    }

    //--------------------------------------------------------------------------
    // Representation of the chessboard

    this.toString = function() {
        var string = ""

        for (var i = 0; i < this.size; i++) {
            if (this.queensCount[i] != 1)
                string += "Queen " + (i + 1) + " not solved\n"
            else
                string += "Queen " + (i + 1) + ": square " + (this.queens[i].indexOf(1) + 1) + "\n"
        }

        return string
    }

    //--------------------------------------------------------------------------
    // Get number of total tries of assignation

    this.steps = function() {
        return this.nsteps
    }
    
    //--------------------------------------------------------------------------
    // Get number of total discards

    this.discards = function() {
        return this.ndiscards
    }
}

module.exports.Chess = Chess
